packages/vite/src/node/server/environment.ts

```diff
@@ -20,26 +20,26 @@ import {
 } from '../optimizer/optimizer'
 import { resolveEnvironmentPlugins } from '../plugin'
 import { ERR_OUTDATED_OPTIMIZED_DEP } from '../constants'
+import { promiseWithResolvers } from '../../shared/utils'
 import { EnvironmentModuleGraph } from './moduleGraph'
 import type { EnvironmentModuleNode } from './moduleGraph'
-import type { HotChannel } from './hmr'
-import { createNoopHotChannel, getShortName, updateModules } from './hmr'
+import type { HotChannel, NormalizedHotChannel } from './hmr'
+import { getShortName, normalizeHotChannel, updateModules } from './hmr'
 import type { TransformResult } from './transformRequest'
 import { transformRequest } from './transformRequest'
 import type { EnvironmentPluginContainer } from './pluginContainer'
 import {
   ERR_CLOSED_SERVER,
   createEnvironmentPluginContainer,
 } from './pluginContainer'
-import type { RemoteEnvironmentTransport } from './environmentTransport'
 import { isWebSocketServer } from './ws'
 
 export interface DevEnvironmentContext {
-  hot: false | HotChannel
+  hot: boolean
+  transport: HotChannel
```

Comment:
Should it be required? 🤔

Reply:
I thought it should be required because the `fetchModule` request has to be handled. But checking it again, `remoteRunner.transport` was optional. What was your intent of that?

Reply:
Because the environment doesn't have to have an associated remote runner, it can be runnerless

Reply:
Done in https://github.com/vitejs/vite/pull/18362/commits/b39895f4d505585431cd673316fdf243df5fb928